publisher/publisher.go

```diff
@@ -116,6 +144,56 @@ func WithInitFunc(f func(conn AMQPConnection) (AMQPChannel, error)) Option {
 	}
 }
 
+func WithNotify(readyCh chan struct{}, unreadyCh chan error) Option {
+	return func(p *Publisher) {
+		p.readyChs = append(p.readyChs, readyCh)
+		p.unreadyChs = append(p.unreadyChs, unreadyCh)
+	}
+}
+
+func (p *Publisher) Notify(readyCh chan struct{}, unreadyCh chan error) (ready <-chan struct{}, unready <-chan error) {
+	if cap(readyCh) == 0 {
+		panic("ready chan is unbuffered")
+	}
+	if cap(unreadyCh) == 0 {
+		panic("unready chan is unbuffered")
+	}
+
+	select {
+	case <-p.NotifyClosed():
+		close(unreadyCh)
+		return readyCh, unreadyCh
+	default:
+	}
+
+	p.mu.Lock()
+	p.readyChs = append(p.readyChs, readyCh)
+	p.unreadyChs = append(p.unreadyChs, unreadyCh)
+	p.mu.Unlock()
+
```

Comment:
tests that cover subscription to notifications at different states:
* dialing -> unready
* connected -> ready
* wait retry -> unready
* paused -> unready
* closed -> unready

Reply:
test cap == 0

Reply:
Make separate `TestNotify` notify with some sub tests